rclrs/src/node/client.rs

```diff
@@ -0,0 +1,257 @@
+
+/// Main class responsible for publishing data to ROS topics
+pub struct Client<T>
+where
+    T: rosidl_runtime_rs::Service,
+{
+    pub(crate) handle: Arc<ClientHandle>,
+    requests: Mutex<HashMap<i64, RequestValue<T::Response>>>,
+    futures: Arc<Mutex<HashMap<i64, oneshot::Sender<T::Response>>>>,
+    sequence_number: AtomicI64,
+}
+
+impl<T> Client<T>
+where
+    T: rosidl_runtime_rs::Service,
+{
+    /// Creates a new client.
+    pub fn new(node: &Node, topic: &str) -> Result<Self, RclrsError>
+    where
+        T: rosidl_runtime_rs::Service,
+    {
+        let mut rcl_client = unsafe { rcl_get_zero_initialized_client() };
+        let type_support = <T as rosidl_runtime_rs::Service>::get_type_support()
+            as *const rosidl_service_type_support_t;
+        let topic_c_string = CString::new(topic).unwrap();
+        let rcl_node = &mut *node.rcl_node_mtx.lock();
+
+        unsafe {
+            let client_options = rcl_client_get_default_options();
+
+            rcl_client_init(
+                &mut rcl_client,
+                rcl_node,
+                type_support,
+                topic_c_string.as_ptr(),
+                &client_options,
+            )
+            .ok()?;
+        }
+
+        let handle = Arc::new(ClientHandle {
+            rcl_client_mtx: Mutex::new(rcl_client),
+            rcl_node_mtx: node.rcl_node_mtx.clone(),
+        });
+
+        Ok(Self {
+            handle,
+            requests: Mutex::new(HashMap::new()),
+            futures: Arc::new(Mutex::new(
+                HashMap::<i64, oneshot::Sender<T::Response>>::new(),
+            )),
+            sequence_number: AtomicI64::new(0),
+        })
+    }
+
+    /// Send a requests with a callback as a parameter.
+    ///
+    /// The [`MessageCow`] trait is implemented by any
+    /// [`Message`] as well as any reference to a `Message`.
+    ///
+    /// The reason for allowing owned messages is that publishing owned messages can be more
+    /// efficient in the case of idiomatic messages[^note].
+    ///
+    /// [^note]: See the [`Message`] trait for an explanation of "idiomatic".
+    ///
+    /// Hence, when a message will not be needed anymore after publishing, pass it by value.
+    /// When a message will be needed again after publishing, pass it by reference, instead of cloning and passing by value.
+    pub fn async_send_request_with_callback<'a, M: MessageCow<'a, T::Request>, F>(
+        &self,
+        message: M,
+        callback: F,
+    ) -> Result<(), RclrsError>
+    where
+        F: FnOnce(&T::Response) + 'static + Send,
+    {
+        let rmw_message = T::Request::into_rmw_message(message.into_cow());
+        let mut sequence_number = self.sequence_number.load(Ordering::SeqCst);
+        let ret = unsafe {
+            rcl_send_request(
+                &*self.handle.lock() as *const _,
+                rmw_message.as_ref() as *const <T::Request as Message>::RmwMsg as *mut _,
+                &mut sequence_number,
```

Comment:
Since this produces a new sequence number, is there any need for us to have the `sequence_number` field at all?

Reply:
Done in 20a09d6ec0798c598f5214c4913b765467d528e1 It used to be the case that the sequence number was not stored in `rcl_client_t` and client libraries needed to keep track of it themselves,, but that was changed.